src/macros.rs

```diff
@@ -33,38 +33,210 @@ macro_rules! scopes {
     }};
 }
 
-/// Reduce boilerplate when inserting new elements in a JSON object.
+/// If there's an optional value, the macro will return its value. Otherwise, a
+/// default value will be returned. This is helpful to handle `$( expr )?`
+/// cases in macros.
 #[doc(hidden)]
 #[macro_export]
-macro_rules! json_insert {
-    ($json:expr, $p1:expr, $p2:expr) => {
-        $json
-            .as_object_mut()
-            .unwrap()
-            .insert($p1.to_string(), json!($p2))
+macro_rules! opt {
+    (, $default:expr) => {
+        $default
     };
+    ($optional:expr, $default:expr) => {
+        $optional
+    };
+}
+
+/// Count items in a list of items within a macro, taken from here:
+/// https://danielkeep.github.io/tlborm/book/blk-counting.html
+#[doc(hidden)]
+#[macro_export]
+macro_rules! replace_expr {
+    ($_t:tt $sub:expr) => {
+        $sub
+    };
+}
+#[doc(hidden)]
+#[macro_export]
+macro_rules! count_items {
+    ($($item:expr),*) => {<[()]>::len(&[$($crate::replace_expr!($item ())),*])};
+}
+
+/// Hack to convert an identifier to a string, including raw identifiers like
+/// `r#type`.
+#[doc(hidden)]
+#[macro_export]
+macro_rules! ident_str {
+    ($name:ident) => {{
+        const NAME: &str = stringify!($name);
+        match &NAME.get(..2) {
+            Some("r#") => &NAME[2..],
+            _ => NAME,
+        }
+    }};
+}
+
+/// Private macro to insert either required or optional fields. Pattern matching
+/// will accordingly pick the branch, and then insert ($key, $val) into $map.
+#[doc(hidden)]
+#[macro_export]
+macro_rules! params_internal {
+    ($map:ident, required, $name:ident, $key:expr, $val:expr) => {
+        $map.insert($key, $val);
+    };
+    ($map:ident, optional, $name:ident, $key:expr, $val:expr) => {
+        // Will only insert when `$name` is not None.
+        if let Some(ref $name) = $name {
```

Comment:
If this statement will only insert when `$name` is not None, would you like to replace `if let Some(ref $name) = $name ` with `if $name.is_some()`. I was confused with this statement in the beginning, it get the reference of `$name`, and what's the next step. I don't get it unill I noticed the comment.